Build and deploy source tarball during release

benesch opened the pull request:
Got around to polishing my free Friday project. See #14031 for the rationale behind this change. In short, this adds a `make archive` target that generates a source tarball that unzips to `src/github.com/cockroachdb/cockroach/...` with submodules vendored. This is exceedingly convenient for Homebrew, and for building from source in general without a Git checkout:

    $ cd workspace
    $ curl https://binaries.cockroachdb.com/cockroach-latest.source.tgz | tar x
    $ export GOPATH="$PWD"
    $ make

Most of the complexity here is refactoring the Makefile to work without a proper Git checkout.

~~**WARNING:** I haven't tested the changes to the deployment process because I don't know how.~~

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14104)
<!-- Reviewable:end -->

benesch committed:
Makefile: suppress .bootstrap error on first build

Prevent Make from printing ".bootstrap: No such file or directory" on
the first build (i.e, when .bootstrap doesn't exist).

benesch committed:
Makefile: add archive target to create source tarball

This allows us to distribute CockroachDB sources without the entire Git
history. Though shallow clones have the same space benefits, this is
still useful: Homebrew, for example, strongly prefers an
upstream-provided source tarball for packaging.

benesch committed:
build: deploy source tarball with binaries during release

mberhault commented on build/push-tagged-aws.sh:
```diff
@@ -16,9 +16,15 @@ set -eux
 ./push-release-tarball.sh cockroach-darwin-10.9-amd64 cockroach-latest.darwin-10.9-amd64
+# Push source tarball (for Homebrew and other package managers that like
+# tarballs). One tagged, one called "latest".
+time aws s3 cp cockroach.tgz "s3://${BUCKET_NAME}/cockroach-latest.source.tgz"
```
the standard (ish) naming is: `src` rather than `source`

mberhault commented on build/push-release-tarball.sh:
```diff
@@ -5,12 +5,11 @@ set -euxo pipefail
 # This file uses `bash` and not `sh` due to the `time` builtin (the external
 # `time` is not available on CircleCI).
 
-BUCKET_NAME="binaries.cockroachdb.com"
```
if you want to leave it configurable, it would be better to use something like:
`BUCKET_NAME="${BUCKET_NAME:-binaries.cockroachdb.com}"`

benesch replied:
I moved this to prevent duplication of the bucket name more than anything else; it wasn't configurable in the original version. Want me to make configurable, though?

mberhault replied:
should be fine since `push-tagged-aws.sh` is the only caller and we have `-eu` on the script.
Another option would be to move all this to two functions inside `push-tagged-aws`, one to do all the tar/gz, the other to just copy.

mberhault commented on build/buildinfo.sh:
```diff
@@ -0,0 +1,50 @@
+    git update-index --add --cacheinfo "100644,$(git hash-object -w --stdin),$1"
+}
+
+if [[ "$1" = tag ]]
```
we tend to use the single-bracket constructions:
`if [ "$1" == "tag"]; then`

mberhault commented on build/buildinfo.sh:
```diff
@@ -0,0 +1,50 @@
+
+if [[ "$1" = tag ]]
+then
+    [[ -f .buildinfo/tag ]] && exec cat .buildinfo/tag
```
I think we should fail when we can't find either the tag or rev, it would be better than ending up with an empty string for either.
The message on failure can mentioned populating `.buildinfo` if someone really wants it to build.

benesch replied:
Make now complains if `buildinfo.sh {tag|rev}` comes back empty. Make makes it incredibly painful to do this any other way.

mberhault commented:
The deployment process is set https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_MergeToMaster which just ends up calling `build/teamcity-release-upload.sh`

We'll want to test this before we need to build the next beta, we'll just have to be careful not to upload the `latest` versions of each thing or we'll overwrite the last beta.

tamird commented:
Reviewed 1 of 1 files at r1, 2 of 4 files at r2, 5 of 5 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[Makefile, line 101 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf79Wi5EX8_XxZPRfwY:-Kf79Wi5EX8_XxZPRfwZ:bcck2kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/Makefile#L101)):*
> ```Makefile
> Unable to determine the build tag and revision.
> If you are compiling without a Git checkout, you need to provide
> a .buildinfo directory with this information. See build/buildinfo.sh
> ```

missing period

---

*[Makefile, line 267 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf79_u_Uipc46Vu0Oc1:-Kf79_u_Uipc46Vu0Oc2:b-j55tgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/Makefile#L267)):*
> ```Makefile
> # manually construct a tree object that includes the vendor submodule at the
> # right path. A future version of Git will likely support this flag.
> $(ARCHIVE): export GIT_ALTERNATE_OBJECT_DIRECTORIES = $(shell git -C vendor rev-parse --git-path objects)
> ```

what's the minimum git version required here?

---

*[Makefile, line 289 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf79Er3WjRPStmbsBmx:-Kf79Er3WjRPStmbsBmy:b-6v4fea) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/Makefile#L289)):*
> ```Makefile
> 
> GIT_DIR := $(shell git rev-parse --git-dir 2> /dev/null)
> ifdef GIT_DIR
> ```

does this actually work the way that this implies? I'd expect GIT_DIR to be defined here, but to have an empty value (in the error case).

Also, what version of git does `--git-dir` require?

---

*[build/buildinfo.sh, line 3 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7AaHRD2v5rsdEppwB:-Kf7AaHRD2v5rsdEppwC:buff3kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/build/buildinfo.sh#L3)):*
> ```Shell
> #!/usr/bin/env bash
> 
> # buildinfo.sh manages tag and revision information in both proper git checkouts
> ```

nit: avoid referring to this script by name. renaming it will rot this comment.

---

*[build/buildinfo.sh, line 17 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7AHZSWcUTPoe66eAj:-Kf7AHZT3-pLx4hpRkOG:bfu9j1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/build/buildinfo.sh#L17)):*
> ```Shell
> # ref`.
> 
> set -eu
> ```

`set -euo pipefail`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 4 files at r2, 2 of 5 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[Makefile, line 95 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7gZb77c6odt93WdfM:-Kf7gZb8YwbjqdQtCyDJ:bnb9rbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L95)):*
> ```Makefile
> export MACOSX_DEPLOYMENT_TARGET=10.9
> 
> BUILDINFO_TAG := $(shell build/buildinfo.sh tag)
> ```

These shell calls will be executed on every call to `make`; the old version only called `ldflags.sh` on `make install`. We should probably move all of this into the `install` target. (not crucial since we already have a lot of stuff that happens on every make, but it's good to not add more)

---

*[Makefile, line 272 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7hrOrjy5PsMhtcMnD:-Kf7hrOrjy5PsMhtcMnE:b-odc5b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L272)):*
> ```Makefile
> 	rm -f $(GIT_INDEX_FILE)
> 	git read-tree HEAD
> 	build/buildinfo.sh inject-archive
> ```

This seems like a lot of cleverness to get git to build the archive for us. Could we do something like `tar czf $(ARCHIVE) .buildinfo $(git ls-files) $(git -C vendor ls-files)`?

---

*[Makefile, line 274 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7jIhZ7zPmLI4_jzCz:-Kf7jIh_-BlGgFtDNR5P:b-hr0kht) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L274)):*
> ```Makefile
> 	build/buildinfo.sh inject-archive
> 	git read-tree -i --prefix=vendor $$(git rev-parse :vendor)
> 	git archive --output=$(ARCHIVE) --prefix=src/github.com/cockroachdb/cockroach/ $$(git write-tree)
> ```

It would be nice if the top level of the source archive included a `Makefile` that set `GOPATH=.` and then delegated to the subdirectory's `Makefile` so users of the source archive don't have to worry as much about `GOPATH` (or dig around for the location of the `Makefile`)

---

*[build/buildinfo.sh, line 4 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7iVRJn7XZAvmlXCqk:-Kf7iVRJn7XZAvmlXCql:by5vyi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/build/buildinfo.sh#L4)):*
> ```Shell
> 
> # buildinfo.sh manages tag and revision information in both proper git checkouts
> # and source tarballs without a .git directory.
> ```

Instead of having two different modes, maybe `buildinfo.sh` should always write to `.buildinfo` and the `Makefile` always reads from `.buildinfo`, and the only difference is that the makefile will automatically re-run `buildinfo.sh` if and only if we're in a git checkout. 

---

*[build/push-tagged-aws.sh, line 28 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7ivozi0eMqhi3ZCTc:-Kf7ivozi0eMqhi3ZCTd:bk6hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/build/push-tagged-aws.sh#L28)):*
> ```Shell
> ./push-release-tarball.sh cockroach-darwin-10.9-amd64 cockroach-latest.darwin-10.9-amd64
> # Push source tarball (for Homebrew and other package managers that like
> # tarballs). One tagged, one called "latest".
> ```

I think I'd also recommend that this tarball be used instead of `git` for our "Build from source" instructions. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

benesch commented:
Ok, this is ready for a re-review! @mberhault I took your advice and merged `push-tagged-aws.sh` and `push-release-tarball.sh`.

---

Review status: 0 of 12 files reviewed at latest revision, 14 unresolved discussions.

---

*[Makefile, line 101 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf79Wi5EX8_XxZPRfwY:-KfFVg4YaZ1WUh46f4hs:bl5m7zz) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/Makefile#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

missing period
</blockquote></details>

Moot now, but FWIW, Make adds a period to the end of any error message.

---

*[Makefile, line 267 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf79_u_Uipc46Vu0Oc1:-KfCL1s_jwsJzdRMVgKB:bmxwc4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/Makefile#L267)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's the minimum git version required here?
</blockquote></details>

The `GIT_ALTERNATE_OBJECT_DIRECTORIES` variable is so old it predates release notes. At least 2005, it seems. https://github.com/git/git/commit/d19938ab6053e3dad75a68a60ef8cad1f378b0e5

---

*[Makefile, line 289 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf79Er3WjRPStmbsBmx:-KfCMlKbO7bMEowvYqc9:b-r4i0sh) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/Makefile#L289)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does this actually work the way that this implies? I'd expect GIT_DIR to be defined here, but to have an empty value (in the error case).

Also, what version of git does `--git-dir` require?
</blockquote></details>

`ifdef` in Make means "if not empty" in this case. ([source](https://www.gnu.org/software/make/manual/html_node/Conditional-Syntax.html)). I think you're right that `ifeq ($(GIT_DIR),)` is clearer, though.

Also, `rev-parse --git-dir` dates back to 2005! https://github.com/git/git/commit/a8783eeb7988a20664dade2379d380ee1a199a10

---

*[Makefile, line 272 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7hrOrjy5PsMhtcMnD:-KfCLG1BKcfiJR-hdvdD:b3wvayr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L272)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems like a lot of cleverness to get git to build the archive for us. Could we do something like `tar czf $(ARCHIVE) .buildinfo $(git ls-files) $(git -C vendor ls-files)`?
</blockquote></details>

I had the same hope; sadly:

$ tar czf out.tgz .buildinfo $(git ls-files) $(git -C vendor ls-files)
-bash: /usr/bin/tar: Argument list too long

Additionally, this (shockingly) turns out to be more portable. There's no way to specify a prefix (`src/github.com/cockroachdb/cockroach`) that works on BSD tar and GNU tar. I'm going to ignore this for now. Plus `git -C vendor ls-files` doesn't include the submodule prefix, so you have to pipe it through awk. It all adds up to approximately the same amount of complexity, no matter which way you go.

---

*[Makefile, line 274 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7jIhZ7zPmLI4_jzCz:-KfFX1iLDbP5-SeGpS51:b-popmdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L274)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It would be nice if the top level of the source archive included a `Makefile` that set `GOPATH=.` and then delegated to the subdirectory's `Makefile` so users of the source archive don't have to worry as much about `GOPATH` (or dig around for the location of the `Makefile`)
</blockquote></details>

Done. See the build/archive directory.

---

*[Makefile, line 293 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfFXuhJYRs5mNWP8hRV:-KfFXuhJYRs5mNWP8hRW:bd2sp5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4f35bad237133a3aa12660681088643ed35cae/Makefile#L293)):*
> ```Makefile
> $(ARCHIVE).tmp: .buildinfo/tag .buildinfo/rev
> 	scripts/ls-files-recursive.sh | $(TAR) -cf $@ -T - $(TAR_XFORM_FLAG),^,$(ARCHIVE_BASE)/src/github.com/cockroachdb/cockroach/, $^
> 	(cd build/archive && $(TAR) -rf ../../$@ --exclude README.md *)
> ```

I'd love to do this in one tar command by specifying multiple `--xform`s, but the macOS `tar` has [a completely crippling bug which makes multiple `--xform`s unusable](https://github.com/libarchive/libarchive/commit/3ea734488052804ff5fd47f6691073ca215e1110).

---

*[build/push-tagged-aws.sh, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf4XYCk9WacJNgdTXpl-r3-29:-KfFX9_e67KjWrkYmHr4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58515955dd51604ad57adcf00fbd26c951644230/build/push-tagged-aws.sh#L29)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

the standard (ish) naming is: `src` rather than `source`
</blockquote></details>

Done.

---

*[build/push-tagged-aws.sh, line 28 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7ivozi0eMqhi3ZCTc:-KfFXA_aWiwu4hGpwJ4s:b-ba7jdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/build/push-tagged-aws.sh#L28)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think I'd also recommend that this tarball be used instead of `git` for our "Build from source" instructions. 
</blockquote></details>

Sounds good to me!

---

*[build/buildinfo.sh, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf4XYCk9WacJNgdTXpj-r3-35:-KfFXWt2FpIg8105XK7j:b-wkmn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/58515955dd51604ad57adcf00fbd26c951644230/build/buildinfo.sh#L35)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

we tend to use the single-bracket constructions:
`if [ "$1" == "tag"]; then`
</blockquote></details>

(Script deleted.)

---

*[build/buildinfo.sh, line 3 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7AaHRD2v5rsdEppwB:-KfFX_9TesJlDdbV83SZ:b-wkmn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/build/buildinfo.sh#L3)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: avoid referring to this script by name. renaming it will rot this comment.
</blockquote></details>

(Script deleted.)

---

*[build/buildinfo.sh, line 17 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7AHZSWcUTPoe66eAj:-KfFXYoJHHLQX0jmCanU:bh20n1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/153c017689f54e72497c503c9a8514ee677eb2fb/build/buildinfo.sh#L17)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`set -euo pipefail`
</blockquote></details>

(Script deleted.)(Script deleted.)

---

*[build/buildinfo.sh, line 4 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7iVRJn7XZAvmlXCqk:-KfFXJxUDZIC72jtzD6r:b-8tt3bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/build/buildinfo.sh#L4)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Instead of having two different modes, maybe `buildinfo.sh` should always write to `.buildinfo` and the `Makefile` always reads from `.buildinfo`, and the only difference is that the makefile will automatically re-run `buildinfo.sh` if and only if we're in a git checkout. 
</blockquote></details>

Good idea, thanks. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-:-KfFlCA6vKdOfUpVGUlg:b-hodkdg)*
<!-- Sent from Reviewable.io -->

mberhault commented on build/teamcity-release-upload.sh:
```diff
@@ -22,6 +22,10 @@ esac
 export VERSION
 echo "Deploying $VERSION..."
 build/builder.sh build/build-static-binaries.sh static-tests.tar.gz
+for archive in cockroach-latest "cockroach-$VERSION"
```
tiny nit: use `${var}` here and on the next line. It makes things clearer, especially when using constructions like `${archive}.src.tgz`

mberhault commented:
LGTM.

You'll need to be extra careful when building the beta release tomorrow. Maybe have someone on hand to quickly review any last-minute changes you need to make (or TBR them).
You'll especially want to fetch/extract all public files and verify the versions/timestamps. Might also be good to tweak the release docs to include verifying the tarball.

tamird commented:
Reviewed 7 of 7 files at r6, 7 of 10 files at r7, 8 of 9 files at r10, 6 of 6 files at r11.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[Makefile, line 293 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGcRAcLV9T2QW33n-n:-KfGcRAcLV9T2QW33n-o:b-sdt1p9) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ca4e9cdba309be9898849ce2880bcda1feff34/Makefile#L293)):*
> ```Makefile
> $(ARCHIVE).tmp: .buildinfo/tag .buildinfo/rev
> 	scripts/ls-files.sh | $(TAR) -cf $@ -T - $(TAR_XFORM_FLAG),^,$(ARCHIVE_BASE)/src/github.com/cockroachdb/cockroach/, $^
> 	(cd build/archive && $(TAR) -rf ../../$@ $(TAR_XFORM_FLAG),^,$(ARCHIVE_BASE)/, --exclude README.md *)
> ```

does this belong in the previous commit? ditto below

---

*[build/build-osx.sh, line 10 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGbbGHNcgI1ijvV483:-KfGbbGIXdL2NOQXOxYs:b955xfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/427121eb951f558d9ea8baa079bcf18e7dad179e/build/build-osx.sh#L10)):*
> ```Shell
> # OSX 10.9 is the most recent version available at time of writing.
> # If changing the OS/arch target, adjust the filename in push-aws.sh.
> make xgo-build GOFLAGS="--image cockroachdb/xgo:20170218 --targets=darwin-10.9/amd64" TYPE=release
> ```

single quote here since you're not interpolating.

---

*[build/archive/README.md, line 1 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGc8iRE1pBX7Ur-JKe:-KfGc8iRE1pBX7Ur-JKf:b-34mf2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/427121eb951f558d9ea8baa079bcf18e7dad179e/build/archive/README.md#L1)):*
> ```Markdown
> # build/archive
> ```

why do we have both this and README?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

benesch commented:
You bet. @tamird, you down? I tested this locally by swapping out the BUCKET_NAME and it all went off with out a hitch.

---

Review status: 0 of 12 files reviewed at latest revision, 13 unresolved discussions.

---

*[Makefile, line 95 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7gZb77c6odt93WdfM:-KfINT8X_mv_nShtYHZh:bv3bxlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L95)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These shell calls will be executed on every call to `make`; the old version only called `ldflags.sh` on `make install`. We should probably move all of this into the `install` target. (not crucial since we already have a lot of stuff that happens on every make, but it's good to not add more)
</blockquote></details>

Should no longer be an issue!

---

*[Makefile, line 272 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf7hrOrjy5PsMhtcMnD:-KfINW9AWW7uz6AEYG7R:b7a0mxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f86596a6becc79908af55828f2f0aa92b1797f1f/Makefile#L272)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I had the same hope; sadly:

$ tar czf out.tgz .buildinfo $(git ls-files) $(git -C vendor ls-files)
-bash: /usr/bin/tar: Argument list too long

Additionally, this (shockingly) turns out to be more portable. There's no way to specify a prefix (`src/github.com/cockroachdb/cockroach`) that works on BSD tar and GNU tar. I'm going to ignore this for now. Plus `git -C vendor ls-files` doesn't include the submodule prefix, so you have to pipe it through awk. It all adds up to approximately the same amount of complexity, no matter which way you go.
</blockquote></details>

I've replaced this with a somewhat-simpler tar command.

---

*[Makefile, line 293 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGcRAcLV9T2QW33n-n:-KfI7scYLOg3lm7WiXz1:b-rr102w) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ca4e9cdba309be9898849ce2880bcda1feff34/Makefile#L293)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does this belong in the previous commit? ditto below
</blockquote></details>

@#% thanks. Fixed.

---

*[build/build-osx.sh, line 10 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGbbGHNcgI1ijvV483:-KfINdqocsUQ9jNlI2JI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/427121eb951f558d9ea8baa079bcf18e7dad179e/build/build-osx.sh#L10)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

single quote here since you're not interpolating.
</blockquote></details>

Done.

---

*[build/teamcity-release-upload.sh, line 25 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-Kf4XYCk9WacJNgdTXpm-r11-25:-KfI7aSBzuOL1AMq7Roh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ca4e9cdba309be9898849ce2880bcda1feff34/build/teamcity-release-upload.sh#L25)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

tiny nit: use `${var}` here and on the next line. It makes things clearer, especially when using constructions like `${archive}.src.tgz`
</blockquote></details>

Done.

---

*[build/archive/README.md, line 1 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGc8iRE1pBX7Ur-JKe:-KfI7b82FWGPM6dJVrPb:b-wqtpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/427121eb951f558d9ea8baa079bcf18e7dad179e/build/archive/README.md#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why do we have both this and README?
</blockquote></details>

Just as an explanation for someone who stumbles across the build/archive directory. Happy to remove it if you think it's clear enough.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-:-KfINLuPoYZYEMy2b0sg:b-5vgx9o)*
<!-- Sent from Reviewable.io -->

tamird commented:
Yup, sounds good.

---

Reviewed 1 of 6 files at r11, 12 of 12 files at r12, 9 of 10 files at r13, 5 of 5 files at r14.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[build/archive/README.md, line 1 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGc8iRE1pBX7Ur-JKe:-KfIPTLf_SGpbphde8_j:b-xhic72) ([raw file](https://github.com/cockroachdb/cockroach/blob/427121eb951f558d9ea8baa079bcf18e7dad179e/build/archive/README.md#L1)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Just as an explanation for someone who stumbles across the build/archive directory. Happy to remove it if you think it's clear enough.
</blockquote></details>

No, I think it's worthwhile, but why not just have one?

---

*[pkg/build/info.go, line 37 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfIPZmiamYewo6cIo5l:-KfIPZmiamYewo6cIo5m:blqbvxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e66224af33c4287074fc7f0c296cbb289989fce/pkg/build/info.go#L37)):*
> ```Go
> 
> // TimeFormat is the reference format for build.Time. Make sure it stays in sync
> // with the string passed by the linker in the root Makefile.
> ```

while you're here: s/by/to/

---

*[scripts/ls-files-recursive.sh, line 5 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfIPeue8ShyZkkWLKX0:-KfIPeue8ShyZkkWLKX1:bh1f39l) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e66224af33c4287074fc7f0c296cbb289989fce/scripts/ls-files-recursive.sh#L5)):*
> ```Shell
> # Recursively list all files in this Git checkout, including files in
> # submodules.
> # TODO(benesch): Replace this with `git ls-files --recurse-submodules` once Git
> ```

nit: this TODO probably wants to be in the caller, which might have more context (i.e. the caller might have context like "i need to support earlier git versions")

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-:-KfIPGP92zC4t8hs18KK:b-bjg5qg)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[build/archive/README.md, line 1 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfGc8iRE1pBX7Ur-JKe:-KfIWKT2_0cjASZOkxqR:bt7irtq) ([raw file](https://github.com/cockroachdb/cockroach/blob/427121eb951f558d9ea8baa079bcf18e7dad179e/build/archive/README.md#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

No, I think it's worthwhile, but why not just have one?
</blockquote></details>

README is copied to the root of the tar archive and is intended for public consumption; README.md is here to explain why this directory exists to developers. If it's confusing already, that's not a good sign!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 6 of 12 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/build/info.go, line 37 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfIPZmiamYewo6cIo5l:-KfIzglAV1TWEi1TZMDd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e66224af33c4287074fc7f0c296cbb289989fce/pkg/build/info.go#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

while you're here: s/by/to/
</blockquote></details>

Done.

---

*[scripts/ls-files.sh, line 5 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfIPeue8ShyZkkWLKX0:-KfIzhG8pkgwnm_TJH8o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e66224af33c4287074fc7f0c296cbb289989fce/scripts/ls-files.sh#L5)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: this TODO probably wants to be in the caller, which might have more context (i.e. the caller might have context like "i need to support earlier git versions")
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 6 files at r11, 1 of 5 files at r14, 6 of 7 files at r15, 5 of 5 files at r16.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

benesch commented:
TFTRs! I'll update the release docs after our release tomorrow. I want to iron out any kinks first.

bdarnell commented:
Review status: 9 of 12 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[build/archive/README.md, line 3 at r18](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104#-KfK6xRsl9IUjqm-SfR3:-KfK6xRtGucCL0bqjKru:b-7kma2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b156123f506f3af6f77479d6945fa7c82edb2f9/build/archive/README.md#L3)):*
> ```Markdown
> # build/archive
> 
> Files in this directory that end in `.archive` are copied to the root of any
> ```

This comment doesn't appear to match the final implementation.

My color for the bikeshed: Put this file in `build/archive/README.md` and the rest in `build/archive/contents/`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14104)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Looks like release builds are now being tagged with `<sha>-dirty` which gets propagated into `cockroach --version`.
This is meant to indicate a manual build from unmerged changes.

jseldess commented:
@benesch, are there implications for the build from source instructions in our [docs](install-cockroachdb.html) and in the [CONTRIBUTING.md](https://github.com/cockroachdb/cockroach/blob/master/CONTRIBUTING.md) file here?

benesch commented:
@mberhault, looking into it.

@jseldess, we may want to switch the build-from-source instructions to use the new tarball instead of the Git repository, but it's not urgent. (In fact, it's not even possible until the next release.) The old method will still work. I'll file a TODO issue, though!
